bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -220,53 +230,89 @@ protected void checkedStart(Listener<RespT> responseListener, Metadata headers)
    *
    * @throws IOException
    */
-  public void asyncRefresh() throws IOException {
+  ListenableFuture<HeaderCacheElement> asyncRefresh() {
     if (isAppEngine) {
-      syncRefresh();
-    } else if (canRefresh()) {
-      executor.execute(new Runnable() {
-        @Override
-        public void run() {
-          doRefresh();
-        }
-      });
+      SettableFuture<HeaderCacheElement> f = SettableFuture.create();
+      try {
+        f.set(syncRefresh());
+      } catch (Exception e) {
+        f.setException(e);
+      }
+      return f;
     }
-  }
 
-  private boolean canRefresh() {
-    return !isRefreshing.get() && getCacheState(this.headerCache.get()) != CacheState.Good;
+    synchronized (this.isRefreshing) {
+      if (!this.isRefreshing.compareAndSet(false, true)) {
+        this.futureToken = executor.submit(new Callable<HeaderCacheElement>() {
+          @Override
+          public HeaderCacheElement call() throws Exception {
+            HeaderCacheElement newToken = refreshCredentialsWithRetry();
+
+            synchronized (RefreshingOAuth2CredentialsInterceptor.this) {
+              headerCache.set(newToken);
+              futureToken = null;
+              isRefreshing.set(false);
+            }
+            return newToken;
+          }
+        });
+      }
+      return this.futureToken;
+    }
   }
 
   /**
    * <p>syncRefresh.</p>
    *
    * @throws java.io.IOException if any.
    */
-  public void syncRefresh() throws IOException {
-    synchronized (isRefreshing) {
-      if (!isRefreshing.get()) {
-        doRefresh();
-      } else {
-        while (isRefreshing.get() && getCacheState(this.headerCache.get()) != CacheState.Good) {
-          try {
-            isRefreshing.wait(250);
-          } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            throw new IOException(e);
-          }
-        }
+  HeaderCacheElement syncRefresh() throws IOException {
+    ListenableFuture<HeaderCacheElement> readerFuture = null;
+    SettableFuture<HeaderCacheElement> writerFuture = null;
+
+    // Check if another thread is already refreshing and set the appropriate future
+    synchronized (this.isRefreshing) {
+      if (this.isRefreshing.get()) {
+        readerFuture = this.futureToken;
+      }
+      else {
+        writerFuture = SettableFuture.create();
```

Comment:
I couldnt figure out a way to share the logic without causing sync stale refreshes from blocking reads. I'm planning on proposing eliminating the sync  path altogether and just wait on the async future for expired tokens in a follow up CL.